docs/grfn_spec.md

```diff
@@ -11,96 +20,189 @@ associated extracted comments, links to natural language text (the
-a variable name includes an \<enclosing\_context\>, for now ignore
-that \<enclosing\_context\>.*
+#### Path strings
 
-A variable name will be a string:
+It will be convenient to be able to express `<scope_path>`s and `<namespace_path>`s using single strings within GrFN (particularly when building an identifier string). For this we introduce a special string notation in which the string names that make up a path are expressed in order but separated by periods. These representations will be referred to as the `<scope_path_string>` and `<namespace_path_string>`, respectively. The string representations of the `<scope_path>` and `<namespace_path>` examples above would be:
 
-    <variable_name> ::= <string>
+- Example `<scope_path_string>`:
+    ```
+    "foo.loop$1.loop$1"
+    ```
+
+- Example `<namespace_path_string>`:
+	```
+	"foo.bar.baz"
+	```
+
+### Identifier string
+
+Identifiers are uniquely defined by their `base_name`, `scope`, and `namespace`. It will be convenient to refer to any identifier using a single string, outside of the identifier specification declaration. We define an `<identifier_string>` by combining the `namespace`, `scope` and `base_name` (in that order) within a single string by separating the `<namespace_path_string>`, `<scope_path_string>` and `<base_name>` by double-colons:
+
+    <identifier_string> ::= "<namespace_path_string>::<scope_path_string>::<base_name>"
+
+### Identifier gensym
+
+One of the outputs of program analysis is a functionally equivalent version of the original source code (e.g., Fortran) and lambda functions, both expressed in the Python (as the intermediate target language). All identifiers in the output Python must match identifiers in GrFN. Since capturing the semantics (particularly the namespace and scope context) results in a representation that does not appear to be consistently expressible in legal Python output, we will use `gensym`s that may be represented (generally more compactly) as legal Python names and associated uniquely with identifiers.
+
+FUTURE: create a unique hashing function between `gensym`s and identifiers.
+
+FOR NOW: Generate Python names that start with a letter followed by a unique integer. The letter could be \'g\' for a generic gensym, or \'v\' to indicate a variable identifier and \'f\' to indicate a function identifier.
+
+Each identifier will be associated one-to-one with a unique `<gensym>`.
 
-A top level source variable named ABSORPTION would then simply have the
-`<source_variable_name>` as the string making up the
-\<variable\_name\>:
+### Identifier specification
 
-    <variable_name> ::= "ABSORPTION"
+Each identifier within a GrFN specification will have a single `<identifier_spec>` declaration, and `<identifier_string>`s are then used to denote uses of the identifier elsewhere in the GrFN spec (e.g., in variable and function specs). An identifier will be declared in the GrFN spec JSON by the following attribute-value list:
```

Comment:
> and `<identifier_string>`s are then used to denote uses of the identifier elsewhere in the GrFN spec (e.g., in variable and function specs).

This portion seems out of place. This section is supposed to be about the identifier specification. I think we should move this important piece of information to the end of the Identifier string section, so that we capture the definition and use case of the identifier string object in one section.

Reply:
Agreed, I have cut this, and slightly reworded the description in the above Identifier string section to:

>It will be convenient to refer unambiguously to any identifier using a single string, outside of the identifier specification declaration (defined below).

and

>`<identifier_string>`s will be used to denote identifiers as they are used in variable and function specifications (described below).